File: Cloude/Cloude/UI/SlashCommand.swift
import SwiftUI
import CloudeShared

struct SlashCommand {
    let name: String
    let description: String
    let icon: String
    let isSkill: Bool
    let resolvesTo: String?
    let parameters: [SkillParameter]

    init(name: String, description: String, icon: String, isSkill: Bool = false, resolvesTo: String? = nil, parameters: [SkillParameter] = []) {
        self.name = name
        self.description = description
        self.icon = icon
        self.isSkill = isSkill
        self.resolvesTo = resolvesTo
        self.parameters = parameters
    }

    var hasParameters: Bool { !parameters.isEmpty }

    static func fromSkill(_ skill: Skill) -> [SlashCommand] {
        let icon = skill.icon ?? "hammer.circle"
        var commands = [SlashCommand(name: skill.name, description: skill.description, icon: icon, isSkill: true, parameters: skill.parameters)]
        for alias in skill.aliases {
            commands.append(SlashCommand(name: alias, description: skill.description, icon: icon, isSkill: true, resolvesTo: skill.name, parameters: skill.parameters))
        }
        return commands
    }
}

let builtInCommands: [SlashCommand] = [
    SlashCommand(name: "compact", description: "Compress conversation context", icon: "arrow.triangle.2.circlepath"),
    SlashCommand(name: "context", description: "Show token usage", icon: "chart.pie"),
    SlashCommand(name: "cost", description: "Show usage stats", icon: "dollarsign.circle"),
    SlashCommand(name: "usage", description: "Usage statistics", icon: "chart.bar.fill"),
]

File: android/app/src/main/java/com/cloude/app/Models/SlashCommand.kt
package com.cloude.app.Models

data class SlashCommand(
    val name: String,
    val description: String,
    val isSkill: Boolean,
    val resolvesTo: String? = null,
    val hasParameters: Boolean = false
) {
    companion object {
        val builtIn = listOf(
            SlashCommand("compact", "Compact conversation context", false),
            SlashCommand("context", "Show context window usage", false),
            SlashCommand("cost", "Show session cost", false)
        )

        fun fromSkills(skills: List<Skill>): List<SlashCommand> {
            val commands = mutableListOf<SlashCommand>()
            skills.filter { it.userInvocable }.forEach { skill ->
                commands.add(
                    SlashCommand(
                        name = skill.name,
                        description = skill.description ?: "",
                        isSkill = true,
                        hasParameters = skill.parameters.isNotEmpty()
                    )
                )
                skill.aliases.forEach { alias ->
                    commands.add(
                        SlashCommand(
                            name = alias,
                            description = skill.description ?: "",
                            isSkill = true,
                            resolvesTo = skill.name,
                            hasParameters = skill.parameters.isNotEmpty()
                        )
                    )
                }
            }
            return commands
        }

        fun allCommands(skills: List<Skill>): List<SlashCommand> =
            builtIn + fromSkills(skills)

        fun filtered(query: String, skills: List<Skill>): List<SlashCommand> {
            val all = allCommands(skills)
            if (query.isEmpty()) return all.filter { it.resolvesTo == null }
            val lower = query.lowercase()
            val exact = all.firstOrNull { it.name.lowercase() == lower }
            if (exact != null) return listOf(exact)
            return all.filter { it.name.lowercase().startsWith(lower) }
        }
    }
}
